File: app/src/main/java/ru/e6atb/chat/MessageAuthorResolver.java
package ru.e6atb.chat;

final class MessageAuthorResolver {
	private MessageAuthorResolver() {
	}

	static MiniTaLib.User resolve(MiniTaLib.Message message, MiniTaLib.User currentRoom) {
		if (message == null) return null;
		MiniTaLib.User author = message.from;
		author = preferMatchingMetadata(author, message.to);
		author = preferMatchingMetadata(author, currentRoom);
		return author;
	}

	private static MiniTaLib.User preferMatchingMetadata(MiniTaLib.User current, MiniTaLib.User candidate) {
		if (candidate == null) return current;
		if (current == null) return candidate;
		if (!sameIdentity(current, candidate)) return current;
		return metadataScore(candidate) > metadataScore(current) ? candidate : current;
	}

	private static boolean sameIdentity(MiniTaLib.User left, MiniTaLib.User right) {
		if (left.id != null && left.id.length() > 0 && left.id.equals(right.id)) return true;
		return left.login != null && left.login.length() > 0 && left.login.equals(right.login);
	}

	private static int metadataScore(MiniTaLib.User user) {
		int score = 0;
		if (user.nick != null && user.nick.length() > 0) score += 8;
		if (user.login != null && user.login.length() > 0) score += 4;
		if (user.roomKind != null && user.roomKind.length() > 0) score += 2;
		if (user.id != null && user.id.length() > 0) score += 1;
		return score;
	}
}

File: app/src/main/kotlin/ru/e6atb/chat/MessageAuthorResolver.kt
@file:Suppress("EXPOSED_FUNCTION_RETURN_TYPE", "EXPOSED_PARAMETER_TYPE")

package ru.e6atb.chat

object MessageAuthorResolver {
    @JvmStatic fun resolve(message: MST5.Message?, currentRoom: MST5.User?): MST5.User? {
        message ?: return null
        return preferMatchingMetadata(preferMatchingMetadata(message.from, message.to), currentRoom)
    }

    private fun preferMatchingMetadata(current: MST5.User?, candidate: MST5.User?): MST5.User? {
        candidate ?: return current
        current ?: return candidate
        return if (sameIdentity(current, candidate) && metadataScore(candidate) > metadataScore(current)) candidate else current
    }

    private fun sameIdentity(left: MST5.User, right: MST5.User) =
        (!left.id.isNullOrEmpty() && left.id == right.id) || (!left.login.isNullOrEmpty() && left.login == right.login)

    private fun metadataScore(user: MST5.User): Int =
        (if (user.nick.isNullOrEmpty()) 0 else 8) + (if (user.login.isNullOrEmpty()) 0 else 4) +
            (if (user.roomKind.isNullOrEmpty()) 0 else 2) + (if (user.id.isNullOrEmpty()) 0 else 1)
}
